src/main/resources/jupyter/google_sign_in.js

```diff
@@ -12,28 +12,46 @@ require.config({
 // TEMPLATED CODE
 // Leonardo has logic to find/replace templated values in the format $(...).
 // This will be replaced with the real Google Client ID before uploading to the notebook server.
-var clientId = $(googleClientId);
+var loginHint = $(googleLoginHint);
+var allowedOrigins = $(allowedOrigins);
 
-require(['gapi'], function(gapi) {
-    gapi.load('auth2', function() {
-        gapi.auth2.init({
-            client_id: clientId,
-            scope: 'email profile openid',
-        }).then(function() {
-            auth2 = gapi.auth2.getAuthInstance();
-            auth2.currentUser.listen(function(user) {
-                authResponse = user.getAuthResponse();
-                set_cookie(authResponse.access_token, authResponse.expires_in);
-            });
-            if (auth2.isSignedIn.get() == false) {
-                auth2.signIn();
+function receive(event) {
+    if (allowedOrigins.indexOf(event.origin) == -1)
+        return;
+    if (event.source !== window.opener)
+        return;
+    if (event.data.type !== 'bootstrap-auth.response')
+        return;
+    startTimer(event.data.body.googleClientId);
+}
+
+function startTimer(googleClientId) {
+    require(['gapi'], function(gapi) {
+        gapi.load('auth2', function() {
+            function doAuth() {
+                gapi.auth2.authorize({
+                    'client_id': googleClientId,
+                    'scope': 'openid profile email',
+                    'login_hint': loginHint,
+                    'prompt': 'none'
+                }, function(result) {
+                    if (result.error) {
+                        return;
+                    }
+                    set_cookie(result.access_token, result.expires_in);
+                });
             }
+            // refresh token every 2 minutes
+            setInterval(doAuth, 120000);
         });
     });
-});
+}
 
 function set_cookie(token, expires_in) {
     var expiresDate = new Date();
     expiresDate.setSeconds(expiresDate.getSeconds() + expires_in);
-    document.cookie = "FCtoken="+token+";expires="+expiresDate.toUTCString();
-}
\ No newline at end of file
+    document.cookie = "FCtoken="+token+";secure;expires="+expiresDate.toUTCString()+";path=/";
+}
+
+window.addEventListener('message', receive);
+window.opener.postMessage('bootstrap-auth.request', '*');
```

Comment:
* because the extension doesn't necessarily know the opener of the notebook (WB or AoU). The WB/AoU side should validate the origin is the Leo URL, and the Leo side validates the postMessage origin against a known list (see above).